### PR TITLE
- Create model 'QuestionAnalytics' with serializer & util files

### PR DESCRIPTION
**Pull request by sirusw**
- Create model 'QuestionAnalytics'
- Create util file init_submission_analytics for initializing SubmissionAnalytics
- Create util file init_question_analytics for initializing QuestionAnalytics
- Create serializer for SubmissionAnalytics and QuestionAnalytics

### CODE REVIEW

File: analytics/migrations/0001_initial.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
-# Generated by Django 3.0.14 on 2021-11-17 22:12
```
Comment (keyvankhademi) [inline analytics/migrations/0001_initial.py:1]: Don't change the existing migrations. When you change the database you need to add migrations.

File: analytics/models.py [hunk 2]
```diff
@@ -30,4 +35,56 @@ class submission_analytics(models.Model):
...
+    vocab = models.DecimalField(max_digits=8, decimal_places=2, default=0)
+    size = models.DecimalField(max_digits=8, decimal_places=2, default=0)
+    vol = models.DecimalField(max_digits=8, decimal_places=2, default=0)
+    difficulty = models.DecimalField(max_digits=8, decimal_places=2, default=0)
+    effort = models.DecimalField(max_digits=8, decimal_places=2, default=0)
+    error = models.DecimalField(max_digits=8, decimal_places=2, default=0)
+    test_time = models.DecimalField(max_digits=8, decimal_places=2, default=0)
+
+
+# class EventAnalytics(models.Model):
```
Comment (keyvankhademi) [inline analytics/models.py:71]: Don't include commented code in the PRs.

File: analytics/tests.py [hunk 3]
```diff
@@ -1,3 +1,203 @@
...
+            event=self.event
+        )
+
+    def test_ensure_db(self):
+        user = MyUser.objects.get(username='test_user2')
+        self.uqj = self.user.question_junctions.filter(question__answer='a').first()
+        create_mcq_submission(self.uqj, '')
+        submission = Submission.objects.first()
+        sub_analytics = SubmissionAnalytics.objects.create(submission_type='mcq', uqj=self.uqj, submission=submission,
+                                                            question=Question.objects.filter(title='title').first(),
```
Comment (keyvankhademi) [inline analytics/tests.py:147]: format this properly.

File: analytics/tests.py [hunk 4]
```diff
@@ -1,3 +1,203 @@
...
+                                                            user_id=user.id, first_name=
+                                                            user.first_name, last_name=user.last_name, ans='a')
+        try:
+            sub_analytics1 = SubmissionAnalytics.objects.get(submission=submission)
+        except SubmissionAnalytics.DoesNotExist:
+            SubmissionAnalytics.objects.create(submission_type='mcq', uqj=self.uqj,
+                                                submission=submission,
+                                                question=Question.objects.get(title='title'),
```
Comment (keyvankhademi) [inline analytics/tests.py:155]: format this properly.

File: api/serializers/submission_analytics.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+class QuestionAnalyticsSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = QuestionAnalytics
+        fields = '__all__'
+
+        depth = 1
+#
+# class EventAnalyticsSerializer(serializers.ModelSerializer):
+#     class Meta:
```
Comment (keyvankhademi) [inline api/serializers/submission_analytics.py:22]: Don't include commented code.

File: analytics/utils/init_question_metrics.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            if wrong_answers[key] > max:
+                wrong_ans = key
+                max = wrong_answers[key]
+
+        try:
+            question_metrics = QuestionAnalytics.objects.get(question=question)
+        except QuestionAnalytics.DoesNotExist:
+            question_metrics = QuestionAnalytics.objects.create(number_submission=number_submission,
+                                                                question=question, event=event,
+                                                                avg_grade=avg_grade/number_submission,
```
Comment (keyvankhademi) [inline analytics/utils/init_question_metrics.py:93]: proper format.

File: analytics/utils/init_analytics.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+        question_obj = Question.objects.get(pk=question.pk)
+        user = mcq_sub_obj.uqj.user
+        user_obj = MyUser.objects.get(pk=user.pk)
+
+        try:
+            sub_analytics = SubmissionAnalytics.objects.get(submission=submission)
+        except SubmissionAnalytics.DoesNotExist:
+            sub_analytics = SubmissionAnalytics.objects.create(submission_type="mcq", uqj=uqj, submission=submission,
+                                                               question=question_obj, event=java_sub_obj.question.event,
+                                                               user_id=user_id,
```
Comment (keyvankhademi) [inline analytics/utils/init_analytics.py:278]: proper format.

File: analytics/utils/init_analytics.py [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+        vol = size * math.log2(vocab)
+        difficulty = unique_operator / 2 + operand / unique_operand
+        effort = vol * difficulty
+        error = vol / 3000
+        test_time = effort / 18
+
+        try:
+            sub_analytics = SubmissionAnalytics.objects.get(submission=submission)
+        except SubmissionAnalytics.DoesNotExist:
+            sub_analytics = SubmissionAnalytics.objects.create(submission_type="parsons", uqj=uqj,
```
Comment (keyvankhademi) [inline analytics/utils/init_analytics.py:248]: proper format

File: analytics/utils/init_analytics.py [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
+        difficulty = unique_operator / 2 + operand / unique_operand
+        effort = vol * difficulty
+        error = vol / 3000
+        test_time = effort / 18
+
+        try:
+            sub_analytics = SubmissionAnalytics.objects.get(submission=submission)
+        except SubmissionAnalytics.DoesNotExist:
+            sub_analytics = SubmissionAnalytics.objects.create(submission_type="java", uqj=uqj, submission=submission,
+                                                               question=question_obj, event=java_sub_obj.question.event,
```
Comment (keyvankhademi) [inline analytics/utils/init_analytics.py:207]: proper format.

File: analytics/views.py [hunk 10]
```diff
@@ -1,9 +1,11 @@
...
 from django.http import HttpResponse
-from .utils import get_submission
+from .utils import init_analytics, init_question_metrics
 
 
 def analysis(request):
-    get_submission.submission_analytics()
+    init_analytics.analytics()
+    init_question_metrics.init()
+    # init_event_metrics.init()
```
Comment (keyvankhademi) [inline analytics/views.py:9]: dont' include comments.

File: api/urls.py [hunk 11]
```diff
@@ -37,6 +38,9 @@
 router.register(r'token-use', TokenUseViewSet, basename='token-use')
 router.register(r'schema', SchemaViewSet, basename='schema')
 router.register(r'difficulty', DifficultyViewSet, basename='difficulty')
+router.register(r'submission-analytics', SubmissionAnalyticsViewSet, basename='submission-analytics')
+router.register(r'question-analytics', QuestionAnalyticsViewSet, basename='question-analytics')
+# router.register(r'event-analytics', EventAnalyticsViewSet, basename='analytics-analytics')
```
Comment (keyvankhademi) [inline api/urls.py:43]: commented code

File: api/views/submission_analytics.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+class QuestionAnalyticsViewSet(viewsets.ReadOnlyModelViewSet):
+    queryset = QuestionAnalytics.objects.all()
+    serializer_class = QuestionAnalyticsSerializer
+    permission_classes = [TeacherAccessPermission]
+    pagination_class = BasePagination
+    filter_class = QuestionAnalyticsFilterSet
+    filter_backends = [DjangoFilterBackend,]
+
+
+# class EventAnalyticsFilterSet(FilterSet):
```
Comment (keyvankhademi) [inline api/views/submission_analytics.py:46]: commented code.

**Review by keyvankhademi — CHANGES_REQUESTED** [review]
I don't think the way you are creating analytics is practical when we have a lot of submissions.
Let's have a meeting regarding this.

**Comment by sirusw** [comment]
Changes:
- Create SubmissionAnalytics upon Submission
- remove comments
- remove other features separate from SubmissionAnalytics
- change some model fields of SubmissionAnalytics from foreign key to number to avoid circular import
- properly format files

File: analytics/migrations/0001_initial.py (hunk 1 above)
Comment (sirusw) [reply analytics/migrations/0001_initial.py:1]: Added.

File: analytics/models.py (hunk 2 above)
Comment (sirusw) [reply analytics/models.py]: Removed.

File: analytics/tests.py (hunk 4 above)
Comment (sirusw) [reply analytics/tests.py]: Removed.

File: api/serializers/submission_analytics.py (hunk 5 above)
Comment (sirusw) [reply api/serializers/submission_analytics.py]: Removed.

File: analytics/utils/init_question_metrics.py (hunk 6 above)
Comment (sirusw) [reply analytics/utils/init_question_metrics.py]: Removed file.

File: analytics/utils/init_analytics.py (hunk 7 above)
Comment (sirusw) [reply analytics/utils/init_analytics.py]: Done.

File: analytics/utils/init_analytics.py (hunk 8 above)
Comment (sirusw) [reply analytics/utils/init_analytics.py]: Done.

File: analytics/utils/init_analytics.py (hunk 9 above)
Comment (sirusw) [reply analytics/utils/init_analytics.py]: Done.

File: analytics/views.py (hunk 10 above)
Comment (sirusw) [reply analytics/views.py]: Removed.

File: api/views/submission_analytics.py (hunk 12 above)
Comment (sirusw) [reply api/views/submission_analytics.py]: Removed.

File: api/urls.py (hunk 11 above)
Comment (sirusw) [reply api/urls.py]: Removed.

**Comment by sirusw** [comment]
Sorry for the mess. I accidentally deleted files when resolving conflicts so I created another PR from another branch with fixes, but the code is identical to this branch.

**Pull request closed without merging**